change/delete trigger updates for cipher objects

srinibas-zededa opened the pull request:
- software library based cipher decryption
- refactored the trigger handlers for ciphercontext and controller certificate.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlecipherconfig.go:
```diff
@@ -133,134 +59,196 @@ func parseCipherContextConfig(getconfigCtx *getconfigContext,
+		cipherCtx.KeyExchangeScheme = cfgCipherContext.GetKeyExchangeScheme()
+		cipherCtx.EncryptionScheme = cfgCipherContext.GetEncryptionScheme()
+		cipherCtx.DeviceCertHash = cfgCipherContext.GetDeviceCertHash()
+		cipherCtx.ControllerCertHash = cfgCipherContext.GetControllerCertHash()
```
Why not intialize the fields inside the {} ?

srinibas-zededa replied:
sure, will do.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlecipherconfig.go:
```diff
@@ -133,134 +59,196 @@ func parseCipherContextConfig(getconfigCtx *getconfigContext,
+	cipherBlock := types.CipherBlock{}
+	if cfgCipherBlock == nil {
+		return cipherBlock
+	}
```
Can't be nil; you're creating a struct.

srinibas-zededa replied:
The cfgCipherBlock is coming from controller, as nil. The CipherBlock returns a structure.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlecipherconfig.go:
```diff
@@ -133,134 +59,196 @@ func parseCipherContextConfig(getconfigCtx *getconfigContext,
+	cipherBlock.CipherContextID = cfgCipherBlock.GetCipherContextId()
+	cipherBlock.InitialValue = cfgCipherBlock.GetInitialValue()
+	cipherBlock.CipherData = cfgCipherBlock.GetCipherData()
+	cipherBlock.ClearTextHash = cfgCipherBlock.GetClearTextSha256()
```
Why not set fields inside the {} ?

srinibas-zededa replied:
will do.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlecipherconfig.go:
```diff
@@ -133,134 +59,196 @@ func parseCipherContextConfig(getconfigCtx *getconfigContext,
+	cipherCtx *types.CipherContext) {
+	// get controller cert
+	if len(cipherCtx.ControllerCert) == 0 {
+		cert := getControllerCert(ctx.zedagentCtx,
```
Seems like this can return an empty byte array if it isn't found. Don't you need at least an error check, and potentially a way to trigger fetching updated controller certs in that case?

srinibas-zededa replied:
Have added the error orchestration through config/status separation for both ControllerCert and CipherContext. Done.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlecertconfig.go:
```diff
@@ -0,0 +1,152 @@
+			return certConfig.Cert
+		}
+	}
+	return []byte{}
```
Better to return a []byte, error, or alternatively []byte, bool, where the boolean indicates whether it was found or not.

srinibas-zededa replied:
done.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlecertconfig.go:
```diff
@@ -0,0 +1,152 @@
+			return certBytes
+		}
+	}
+	return []byte{}
```
What do other agents do when they can't read the device cert? I suspect they all log.Error. They might even do more drastical things. But silence like here sounds quite bad.

srinibas-zededa replied:
This is now part of orchestration. The error will be posted to controller.

eriknordmark commented on pkg/pillar/cmd/zedagent/handlecipherconfig.go:
```diff
@@ -133,134 +59,196 @@ func parseCipherContextConfig(getconfigCtx *getconfigContext,
+	// get device cert
+	if len(cipherCtx.DeviceCert) == 0 {
+		cert := getDeviceCert(cipherCtx.HashScheme,
+			cipherCtx.DeviceCertHash)
```
If the deviceCertHash isn't found I suspect you should generate an error, because the controller is clearly confused.

srinibas-zededa replied:
Sure. Done.

eriknordmark commented:
@srinibas-zededa let me know when you have addressed the comments so we can pull this to master,

srinibas-zededa commented:
> @srinibas-zededa let me know when you have addressed the comments so we can pull this to master,
Hi Erik, I have refactored more on the line of config/status model. Wherein the errors, can be captured as they occur. And can be conveyed to controller through Device Info or, similar API.

eriknordmark commented on pkg/pillar/cmd/downloader/syncop.go:
```diff
@@ -322,28 +322,54 @@ func sourceFailureError(ip, ifname, url string, err error) {
 	return cred, nil
 }
+
+func handleCipherBlockError(ctx *downloaderContext,
```
I would be useful to have a comment about this function. I only know what it does because of what Saurabh told me today.

srinibas-zededa replied:
Done.

eriknordmark commented on pkg/pillar/cmd/domainmgr/domainmgr.go:
```diff
@@ -2839,32 +2883,56 @@ func maybeResizeDisk(diskfile string, maxsizebytes uint64) error {
+	config types.DomainConfig) (*string, error) {
+	cipherBlock := new(types.CipherBlockStatus)
+	*cipherBlock = config.CipherBlockStatus
+	cipherBlock.CipherBlockID = getCipherBlockID(cipherBlock.Key(),
```
What is this ID used for? I can't quite understand why you have the AppNum in there.

srinibas-zededa replied:
Removed, the cipherblock will use the config.Key() as ID.

eriknordmark commented on pkg/pillar/devicenetwork/devicenetwork.go:
```diff
@@ -385,30 +385,52 @@ func devPortInstallWifiConfig(ifname string, wconfig types.WirelessConfig) bool
 	return cred, nil
 }
 
+func handleCipherBlockError(status *types.CipherBlockStatus,
```
You have two copies of this. Can it go in the utils package?

eriknordmark replied:
Or as a meber function of type.CipheblockStatus?

srinibas-zededa replied:
Have moved it to utils (cipherutils.go)

eriknordmark commented on pkg/pillar/types/certinfotypes.go:
```diff
@@ -4,20 +4,50 @@
+	//Type     zcert.ZCertType
+	Cert     []byte
+	CertHash []byte
+	ErrorInfo
```
Is there an API to report this to the controller?

srinibas-zededa replied:
A new API will be defined. It will periodically publish the object cipher status and error, if any.

eriknordmark replied:
Please create a story for that.

eriknordmark commented:
Still waiting for test results. No need to include them here but please send them out.
First of all need to test against current controller version, then also test against the controller sending the encrypted blobs.

eriknordmark commented on pkg/pillar/cmd/domainmgr/domainmgr.go:
```diff
@@ -697,6 +731,10 @@ func handleDomainDelete(ctxArg interface{}, key string,
+	// delete the specific cipher block status
+	ctx := ctxArg.(*domainContext)
+	config := configArg.(types.DomainConfig)
+	unpublishCipherBlockStatus(ctx, config.Key())
```
Note that you only publish if you had cloudinit user data, but you unconditionally unpublish. Hence you get some log.Error noise which is undesriable

srinibas-zededa replied:
Have put a check, to not to publish, incase no valid cloud init user data (cipher or, plain)
